### PR TITLE
base-template

### PR DESCRIPTION
**Pull request by bucktower**
Added a base level .eslintrc for javascript, typescript compatible with Node (hapi example) and React (CRA example), and fleshed out the README.

### CODE REVIEW

**Comment by deandalm** [comment]
@reneedudley should we add the web team to this review?

File: js/node/hapi/main.js
Comment (johnny-1111one1) [inline js/node/hapi/main.js]: Is this an example? Why is this in the linting project? I could understand examples that highlight why certain linting choices were made over others, but I'm not sure I see the purpose of why this file is here.

**Review by johnny-1111one1 — COMMENTED** [review]
I'm not sure I understand this PR relative to the purpose of this project. I was expecting a linting file/config that can be included as a library in another project, which established consistent linting rules to be used across all of our projects.

I have not looked through this project, but creating a linting project like the one we include from Dialexa: https://github.com/dialexa/eslint-config-dialexa

And the Ruby version of what we include is here: https://github.com/silvercar/silvercop

**Review by stu-k — COMMENTED** [review]
I was under the assumption this would just be a repo that acts as our importable library for an `.eslint` config

File: js/node/hapi/main.js
Comment (reneemeyer) [reply js/node/hapi/main.js]: This is not needed in this project.

**Review by reneemeyer — CHANGES_REQUESTED** [review]
The initial eslint rules look fine as does the tsconfig, however all of the other files pertaining to create react app and hapi need to be discarded. Also as the ticket states, this should be published to npm so a dev doesn't need to "drop" a file into the project.

**Comment by johnny-1111one1** [comment]
General questions, I'm not sure if it comes from airbnb's default or not, I didn't see specific rules regarding it:

* What are the defaults for how many lines of code a class can be?
* What are the defaults for how many lines of code a method can be?
* What are the defaults for how many parameters can be passed into a method?

**Comment by bucktower** [comment]
Restructured, removed examples, installation of npm package and linting working locally using the following steps:

1. `npx create-react-app yetanotherapp --typescript`
2. `cd yetanotherapp && npm i --save-dev ../js-constable`
3. `vim .eslintrc`
4.
```json
{
  "extends": "./node_modules/@silvercar/js-constable/react-ts"
}
```
5. Open up any `.tsx` file in VSCode and linting errors show

File: rules/js/react/.esmintrc
Comment (stu-k) [inline rules/js/react/.esmintrc]: Is this file supposed to be `.eslintrc`?

File: README.md
Comment (stu-k) [inline README.md]: I may be wrong, but I thought this repo was going to be public. If so, we shouldn't be linking internal channels.

File: README.md
Comment (stu-k) [inline README.md]: Super pedantic, but can we use `const`s in the example instead of `var`.

File: README.md
Comment (stu-k) [inline README.md]: We should probably be using an s3 for this asset, not imgur.

File: README.md
Comment (stu-k) [inline README.md]: Should remove until it's implemented.

File: README.md
Comment (stu-k) [inline README.md]: Should remove until it's implemented.

File: README.md
Comment (reneemeyer) [reply README.md]: We will also get a more fun logo. When ready I can put it on s3

File: rules/js/react/.esmintrc
Comment (reneemeyer) [inline rules/js/react/.esmintrc]: I think we should turn this on, or at the least move it to a warning.

File: README.md
Comment (johnny-1111one1) [reply README.md]: Also not sure that we need this channel. I joined so I could say that the guild channels or eng-team would probably be more appropriate. Not sure we need a whole channel for linting since the more people in a channel that see it, the better.

**Review by johnny-1111one1 — DISMISSED** [review]
Should have hit Request Changes when I left my first comment. I'm not as familiar with the rules in ESLint, but do want to make sure there are sane default values for (they're not referenced in this PR, so I'm hoping there are defaults in the base config we're building on top of):
`max-params`
`max-lines-per-function`

I wish there was a `max-lines-per-class`, but I'm also not sure how JS interprets function vs class (or a React Component), but some combination of `max-lines` and `max-class-per-file` could help mitigate the size and single responsibility of classes.

File: rules/js/react/.esmintrc
Comment (bucktower) [reply rules/js/react/.esmintrc]: Haha yesh

File: README.md
Comment (bucktower) [reply README.md]: Left channel (I'm not an admin so I can't delete it), updated README

**Comment by stu-k** [comment]
@johnny-1111one1 I definitely agree with using the [max params](https://eslint.org/docs/rules/max-params) rule, but as for the length of classes, they can get pretty long while still staying very simple, just by nature of having JSX in them. We could add [other checks (See Related Rules)](https://eslint.org/docs/rules/max-lines#related-rules) to keep them simpler, but I've seen very long classes/functions that don't look out of place.

**Comment by johnny-1111one1** [comment]
@stuartkilgore  The `complexity` and `max-statements` look like they could be useful. My concern in bringing up these rules is not that simple classes can get very long, but by using linting rules to help developers follow OO/SOLID principles, especially Single Responsibility. I want to make it difficult or impossible for someone to write procedures instead of OO classes. Linting rules about size and complexity are very good at that.

I'm not as confident on the React/UI type stuff, but Node should be able to follow the same kind of rules as other backend languages, but can vary the exact number (like Java can be very verbose and need higher thresholds):
* No more than 4 params to any method
* Methods cannot be longer than 10 lines
* Classes cannot be longer than 100 lines

Those kind of rules.
If instead, we need to say:
* Complexity must be less than 15 (is this Cyclomatic complexity?)
* Max statements cannot be more than 10

About React thresholds, I would think that creating components would reduce procedural-type code as class-based components can encapsulate logic.

File: README.md
Comment (bucktower) [reply README.md]: Hosting within repo, as suggested by @brycehendrix on #guild-devops

File: rules/js/react/.esmintrc
Comment (bucktower) [reply rules/js/react/.esmintrc]: Moved no-shadow to error, the rest to warnings, along with some other reevaluations of previously disabled ones.

**Review by bucktower — COMMENTED** [review]
@johnny-1111one1 Attempted to explicitly curb run on code, but do so forgivingly to where it isn't a complete headache when we first drop in linting and rules are actually followed from the get-go. Using just warnings for right now, with more generous limits, and then we can gradually tighten length requirements and get stricter with npm updates. But this is all subjective -- thoughts? Also @stuartkilgore what do you think, since this is definitely geared towards frontend/React at first.

```js
    'max-lines': ['warn', { max: 500, skipBlankLines: true, skipComments: true }],
    'max-classes-per-file': ['warn', 3],
    'max-lines-per-function': ['warn', { max: 75 }],
    'max-params': ['warn', 4],
    complexity: ['warn', 15],
    'max-statements': ['warn', 15],
```

File: rules/common.js
Comment (johnny-1111one1) [inline rules/common.js:34]: Since you can pass in an array, can you specify something like: `['error', { max: 75 }, 'warn', { max: 25 }]`? I can't tell if it's just an array...because (and should be a hash with a severity key), or supports multiple values.

File: rules/common.js
Comment (johnny-1111one1) [inline rules/common.js:30]: Is `1` "warning" or `true`? I see `2` above. And then I see something like `skipComments: true`.

File: rules/common.js
Comment (johnny-1111one1) [inline rules/common.js:35]: Does this need to be a string?

File: rules/common.js
Comment (johnny-1111one1) [inline rules/common.js]: I would make this an error. If someone makes this function: `doSomething(a, b, c, d, e, f, g, h, i, j, k, l, m)`, that's wrong.

**Review by johnny-1111one1 — COMMENTED** [review]
I think the line stuff is pretty lenient and will defer to others as long as we have some sane values. I also don't know enough about the linting rules in general, such as if it's aware of code smells and helps protect against those as well (temporary variables, unused assignment, etc -- I need to familiarize myself with convention), stuff like security, etc.

File: rules/common.js
Comment (bucktower) [reply rules/common.js:30]: `0` == `off`, `1` == `warning`, `2` == `error`

File: rules/common.js
Comment (bucktower) [reply rules/common.js:35]: Haha yeah thats what I thought too but linting flags it as 'unnecessary' 🤷‍♂️work's just fine with CRA test-case with Typescript -- just tested again now

File: rules/common.js
Comment (bucktower) [reply rules/common.js:34]: This is actually a really good idea! I looked into it [here](https://github.com/eslint/eslint/issues/3427) and it isn't a thing in vanilla ESLint, but we could perhaps write an ESLint plugin with warning-error thresholds at some point.

**Comment by reneemeyer** [comment]
@bucktower the only thing I would suggest tightening up is the classes per file. It's best practice in React to have 1 class/1 file. Other than that, I'm ok with this as a first pass!

**Comment by bucktower** [comment]
@reneedudley Figured out how to do it just for React in TS using ESLint `overrides` since in non-React. Review again? 😁

File: README.md
Comment (stu-k) [inline README.md]: Can remove this section.

**Review by stu-k — APPROVED** [review]
Most all comments have been addressed, one additional nitpick in the readme 👍

File: README.md
Comment (GPFAFF) [inline README.md]: Is this indent intentional?

File: rules/common.js
Comment (GPFAFF) [inline rules/common.js]: This probably could be cut in half and even then that is a tad high. I believe the intention is to dial it back, but maybe we make a guesstimate around what the ideal file length should be and then adjust accordingly from there.

File: rules/react.js
Comment (GPFAFF) [inline rules/react.js:38]: 👍

**Review by GPFAFF — COMMENTED** [review]
Great job on this @bucktower

**Comment by deandalm** [comment]
@reneedudley is the plan to add CircleCI later?

**Comment by bucktower** [comment]
> @reneedudley is the plan to add CircleCI later?

@deandalm Yup workin on that right now: https://circleci.com/gh/silvercar/js-constable/tree/circleci

**Review by reneemeyer — APPROVED** [review]

**Comment by deandalm** [comment]
@reneedudley and @bucktower is the plan to integrate this and add CircleCI in another PR? 

If that is the case I can remove the build error check (since this branch doesn't have a CircleCI config) and add it back once this is merged in.

**Comment by bucktower** [comment]
@deandalm Yes this is the case. Going to pull request from circleci branch

**Comment by deandalm** [comment]
@bucktower you can ignore the `Build Error` and merge this in.

**Pull request merged**
